naming/etcd/etcd.go

```diff
@@ -0,0 +1,145 @@
+package etcd
+
+import (
+	"log"
+	"sync"
+
+	etcdcl "github.com/coreos/etcd/client"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/naming"
+)
+
+type kv struct {
+	key, value string
+}
+
+// recvBuffer is an unbounded channel of *kv to record all the pending changes from etcd server.
+type recvBuffer struct {
+	c        chan *kv
+	mu       sync.Mutex
+	stopping bool
+	backlog  []*kv
+}
+
+func newRecvBuffer() *recvBuffer {
+	b := &recvBuffer{
+		c: make(chan *kv, 1),
+	}
+	return b
+}
+
+func (b *recvBuffer) put(r *kv) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.stopping {
+		return
+	}
+	b.backlog = append(b.backlog, r)
+	select {
+	case b.c <- b.backlog[0]:
+		b.backlog = b.backlog[1:]
+	default:
+	}
+}
+
+func (b *recvBuffer) load() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.stopping || len(b.backlog) == 0 {
+		return
+	}
+	select {
+	case b.c <- b.backlog[0]:
+		b.backlog = b.backlog[1:]
+	default:
+	}
+}
+
+func (b *recvBuffer) get() <-chan *kv {
+	return b.c
+}
+
+func (b *recvBuffer) stop() {
```

Comment:
// stop terminates the recvBuffer. After it is called, the recvBuffer is not usable any more.